File: Python/257. Binary Tree Paths.py
'''
回溯
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        path = []
        res = []
        def backtrack(node):
            path.append(str(node.val))
            if not node.left and not node.right:
                res.append('->'.join(path))
            if node.left:
                backtrack(node.left)
            if node.right:
                backtrack(node.right)
            path.pop(-1)
        backtrack(root)
        return res

File: C++/257. Binary Tree Paths.cpp
/*
1.»ØËÝ
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> path;
    vector<string> res;
    void backtrack(TreeNode* node) {
        path.push_back(node->val);
        if (!node->left && !node->right) {
            string sPath;
            for (int num : path) {
                sPath += to_string(num);
                sPath += "->";
            }
            res.push_back(sPath.substr(0, sPath.size() - 2));
        }
        if (node->left) {
            backtrack(node->left);
        }
        if (node->right) {
            backtrack(node->right);
        }
        path.pop_back();
    }
    vector<string> binaryTreePaths(TreeNode* root) {
        path.clear();
        res.clear();
        backtrack(root);
        return res;
    }
};